### PR TITLE
Fix bug in MPIBoundaryCommSetupHelper

### PR DESCRIPTION
**Pull request by majosm**
I found the source of my CI hangs in #133. `MPIBoundaryCommSetupHelper` was doing something not so safe: it was waiting on its send request immediately after receiving, instead of waiting until all the receives are done. This PR tweaks its API to allow waiting on the sends separately.

I think grudge managed to avoid bumping into this because it completes the receives in order (unlike the partition tests here).

:warning:  Should land at the same time as https://github.com/inducer/grudge/pull/63

### CODE REVIEW

File: test/test_partition.py [hunk 1]
```diff
@@ -368,19 +371,24 @@ def _test_mpi_boundary_swap(dim, order, num_groups):
...
+    pending_recvs = connected_parts.copy()
+    while pending_recvs:
+        for i_remote_part in pending_recvs:
+            setup_helper = bdry_setup_helpers[i_remote_part]
+            if setup_helper.is_recv_ready():
+                conn = setup_helper.recv()
                 check_connection(actx, conn)
                 remote_to_local_bdry_conns[i_remote_part] = conn
+                pending_recvs.remove(i_remote_part)
                 break
```
Comment (inducer) [inline test/test_partition.py:387]: I think this ultimately busy-waits, which I'm not a huge fan of. Insert some form of MPI wait instead? `MPI_Waitall` is not ideal, but it'll do, and it's simple. `MPI_Waitsome` if you're feeling fancy.

**Review by inducer — COMMENTED** [review]
Yikes! Thanks for figuring this out. (And sorry for the trouble!) One minor wrinkle, otherwise good to go.

File: test/test_partition.py (hunk 1 above)
Comment (majosm) [reply test/test_partition.py]: The existing `MPIBoundaryCommSetupHelper` design wasn't really compatible with `Wait[all/any/some]`, so I made some changes to it. It's now able to use `Waitsome` under the hood. However it turns out the current release of mpi4py doesn't yet support `Waitsome`... or at least not the lowercase version that works with pickled objects. It's there in the main branch, but it looks like it was added sometime after the latest release (which is surprisingly old -- 2019). In the meantime I made it use `Waitany` as a workaround.

File: meshmode/distributed.py [hunk 2]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+    .. automethod:: complete_some
     """
-    def __init__(self, mpi_comm, actx, local_bdry_conn, i_remote_part,
+    def __init__(self, mpi_comm, actx, connected_parts, local_bdry_conns,
             bdry_grp_factory):
         """
         :arg mpi_comm: A :class:`MPI.Intracomm`
-        :arg i_remote_part: The part number of the remote partition
+        :arg actx: An array context
+        :arg connected_parts: A `set` of remote partitions that are connected to the
```
Comment (inducer) [inline meshmode/distributed.py:169]:
```suggestion
        :arg connected_parts: A :class:`set` of remote partitions that are connected to the
```

File: meshmode/distributed.py [hunk 3]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+            local partition
```
Comment (inducer) [inline meshmode/distributed.py:170]: Are these just the numbers of the connected partitions? If not, what is a "partition"?

Comment (inducer) [inline meshmode/distributed.py:170]: Would `connected_parts` agree with `set(local_bdry_conns)` (i.e. the keys)?

File: meshmode/distributed.py [hunk 4]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+        :arg local_bdry_conns: A `dict` mapping remote partition to
```
Comment (inducer) [inline meshmode/distributed.py:171]:
```suggestion
        :arg local_bdry_conns: A :class:`dict` mapping remote partition to
```

File: meshmode/distributed.py [hunk 5]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+        self._internal_mpi_comm = None
+        self.send_reqs = None
+        self.recv_reqs = None
 
-        self.send_req = None
+    def __enter__(self):
+        self._internal_mpi_comm = self.mpi_comm.Dup()
 
-    def _post_send_boundary_data(self):
+        logger.info("bdry comm rank %d recv begin", self.i_local_part)
```
Comment (inducer) [inline meshmode/distributed.py:191]:
```suggestion
        logger.info("bdry comm rank %d comm begin", self.i_local_part)
```

File: meshmode/distributed.py [hunk 6]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+        self.recv_reqs = []
+        for i_remote_part in self.connected_parts:
+            self.recv_reqs.append(
+                self._internal_mpi_comm.irecv(source=i_remote_part))
 
-        return self.mpi_comm.isend(
-                (self.local_bdry_conn.to_discr.mesh,
+        logger.info("bdry comm rank %d send begin", self.i_local_part)
```
Comment (inducer) [inline meshmode/distributed.py:197]:
```suggestion
```

File: meshmode/distributed.py [hunk 7]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
-        self.send_req = None
+    def __enter__(self):
+        self._internal_mpi_comm = self.mpi_comm.Dup()
 
-    def _post_send_boundary_data(self):
+        logger.info("bdry comm rank %d recv begin", self.i_local_part)
+        self.recv_reqs = []
+        for i_remote_part in self.connected_parts:
+            self.recv_reqs.append(
+                self._internal_mpi_comm.irecv(source=i_remote_part))
```
Comment (inducer) [inline meshmode/distributed.py:195]: Make this a list comprehension.

File: meshmode/distributed.py [hunk 8]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+        for i_remote_part in self.connected_parts:
+            self.send_reqs.append(
+                self._internal_mpi_comm.isend((
+                    self.local_bdry_conns[i_remote_part].to_discr.mesh,
                     make_remote_group_infos(
-                        self.array_context, self.local_bdry_conn)),
-                dest=self.i_remote_part,
-                tag=TAG_SEND_BOUNDARY)
+                        self.array_context, self.local_bdry_conns[i_remote_part])),
+                    dest=i_remote_part))
```
Comment (inducer) [inline meshmode/distributed.py:205]: Make this a list comprehension.

File: meshmode/distributed.py (hunk 5 above)
Comment (inducer) [inline meshmode/distributed.py:191]: Add a corresponding completion message somewhere, maybe?

File: test/test_partition.py [hunk 9]
```diff
@@ -353,35 +353,25 @@ def _test_mpi_boundary_swap(dim, order, num_groups):
...
-    from meshmode.discretization.connection import check_connection
-    while bdry_setup_helpers:
-        for i_remote_part, setup_helper in bdry_setup_helpers.items():
-            if setup_helper.is_setup_ready():
-                assert bdry_setup_helpers.pop(i_remote_part) is setup_helper
-                conn = setup_helper.complete_setup()
+    with MPIBoundaryCommSetupHelper(mpi_comm, actx, connected_parts,
+            local_bdry_conns, bdry_grp_factory=group_factory) as bdry_setup_helper:
+        from meshmode.discretization.connection import check_connection
+        while conns := bdry_setup_helper.complete_some():
```
Comment (inducer) [inline test/test_partition.py:371]: I would love to use walruses, too! Unfortunately, they're not yet in meshmode's compatibility target (3.6).

File: meshmode/distributed.py [hunk 10]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
-        Returns the tuple ``remote_to_local_bdry_conn``
-        where `remote_to_local_bdry_conn` is a
-        :class:`DirectDiscretizationConnection` that gives the connection that
-        performs data exchange across faces from partition `i_remote_part` to the
-        local mesh.
+        Returns a `dict` mapping a subset of remote partitions to remote-to-local
+        boundary connections, where a remote-to-local boundary connection is a
+        :class:`DirectDiscretizationConnection` that performs data exchange across
+        faces from partition `i_remote_part` to the local mesh.
         """
```
Comment (inducer) [inline meshmode/distributed.py:218]: It seems that an empty returned dict indicates that we're done. Is that generally true? If so, document it.

File: meshmode/distributed.py [hunk 11]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+                    self.array_context,
+                    local_bdry_conn=local_bdry_conn,
+                    i_local_part=self.i_local_part,
+                    remote_bdry_discr=local_bdry_conn.to_discr.copy(
+                        actx=self.array_context,
+                        mesh=remote_bdry_mesh,
+                        group_factory=self.bdry_grp_factory),
+                    remote_group_infos=remote_group_infos)
+
+        if not remote_to_local_bdry_conns:
```
Comment (inducer) [inline meshmode/distributed.py:251]: Could we pop the "done" requests out of `recv_reqs`? (probably in reverse order sorted so as to not mess up later indices)

Comment (inducer) [inline meshmode/distributed.py:251]:
```suggestion
        completed = not remote_to_local_bdry_conns
        if completed:
```

**Review by inducer — COMMENTED** [review]
Thanks! I like how this turned out. A few nits to pick, otherwise great.

File: meshmode/distributed.py [hunk 12]
```diff
@@ -155,72 +154,104 @@ def make_remote_group_infos(actx, bdry_conn):
...
+        self.connected_parts = list(connected_parts)
+        self.local_bdry_conns = local_bdry_conns
         self.bdry_grp_factory = bdry_grp_factory
+        self._internal_mpi_comm = None
+        self.send_reqs = None
+        self.recv_reqs = None
 
-        self.send_req = None
+    def __enter__(self):
+        self._internal_mpi_comm = self.mpi_comm.Dup()
```
[the remark this answers is not in the document]
Comment (majosm) [reply meshmode/distributed.py:189]: > Is the idea to keep this communication separate from everything else without having to fix a tag?

Yep.

File: meshmode/distributed.py (hunk 3 above)
Comment (majosm) [reply meshmode/distributed.py:170]: They're just the numbers, yeah.

Comment (majosm) [reply meshmode/distributed.py:170]: Yes they would. Should I get rid of `connected_parts`?

Comment (inducer) [reply meshmode/distributed.py]: Yeah, I'd say so. I don't see why not.

File: meshmode/distributed.py (hunk 11 above)
Comment (inducer) [reply meshmode/distributed.py]: (As discussed: nvm)

**Review by inducer — APPROVED** [review]
Thanks for working on this! I like how it turned out.

**Comment by majosm** [comment]
@inducer So, apparently this is broken.

https://bitbucket.org/mpi4py/mpi4py/issues/65/mpi_err_truncate-message-truncated-when

> The implementation of irecv() for large messages requires users to pass a buffer-like object large enough to receive the pickled stream. This is not documented (as most of mpi4py), and even non-obvious and unpythonic, but if you have some good knowledge about MPI you will will understands its limitations for implementing irecv() for pickled streams.

:eyes:

**Comment by majosm** [comment]
Failing mirgecom CI: https://github.com/illinois-ceesd/mirgecom/pull/316/checks?check_run_id=2335341297

```
Traceback (most recent call last):
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/site-packages/mpi4py/__main__.py", line 7, in <module>
    main()
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/site-packages/mpi4py/run.py", line 196, in main
    run_command_line(args)
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/site-packages/mpi4py/run.py", line 47, in run_command_line
    run_path(sys.argv[0], run_name='__main__')
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "./examples/lump-mpi.py", line 160, in <module>
    main()
  File "/home/runner/work/mirgecom/mirgecom/.miniforge3/envs/testing/lib/python3.8/site-packages/mirgecom-2020.1-py3.8.egg/mirgecom/mpi.py", line 154, in wrapped_func
    func(*args, **kwargs)
  File "./examples/lump-mpi.py", line 102, in main
    discr = EagerDGDiscretization(
  File "/home/runner/work/mirgecom/mirgecom/src/grudge/grudge/discretization.py", line 100, in __init__
    self._set_up_distributed_communication(
  File "/home/runner/work/mirgecom/mirgecom/src/grudge/grudge/discretization.py", line 140, in _set_up_distributed_communication
    conns = bdry_setup_helper.complete_some()
  File "/home/runner/work/mirgecom/mirgecom/src/meshmode/meshmode/distributed.py", line 222, in complete_some
    index, msg = MPI.Request.waitany(self.recv_reqs)
  File "mpi4py/MPI/Request.pyx", line 252, in mpi4py.MPI.Request.waitany
  File "mpi4py/MPI/msgpickle.pxi", line 446, in mpi4py.MPI.PyMPI_waitany
mpi4py.MPI.Exception: Message truncated, error stack:
PMPI_Waitany(282): MPI_Waitany(count=1, req_array=0x7f6dc8c2f7e0, index=0x7fff6e4a66c8, status=0x7fff6e4a66f0) failed
PMPI_Waitany(265): 
do_cts(548)......: Message truncated; 84029 bytes received but buffer size is 32768
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
Example ./examples/lump-mpi.py failed.
```

cc @matthiasdiener

**Comment by inducer** [comment]
Yikes! Interesting. There seems to be an easy fix though: we can pickle outside of mpi4py (so that we know how big the buffers need to be), allreduce (max) that, and then pass the buffer it seems to want. That should address the problem, right?

**Comment by inducer** [comment]
(Just heard @majosm speak at the CS meeting, which brought me back here.) It occurred to me we don't even need to allreduce. Each node can just tell the other how big a buffer is coming during a "first round", and then post the receive with the adequately-sized buffer after that.

**Pull request merged**